### ISSUE TITLE
Should IteratorResult have optional value?

### ISSUE THREAD

**Issue by christyharagan** [Microsoft/TypeScript#8938]
Looking at the ECMA spec for IteratorResult, it seems that ```value``` can be optional (when ```done``` is ```true```)

In the TypeScript definition, this would translate to:

```ts
interface IteratorResult<T> {
    done: boolean;
    value?: T;
}
```

This does create a potentially annoying situation in which, when ``strictNullChecks``` is true, the ```value``` has to be cast to ```T``` (unless the new control-flow based type guards are smart enough to recognise this).

As it is, however, it isn't possible to actually implement an ```Iterator``` with ```strictNullChecks``` because  we can't return an undefined value for ```value```.

What is the correct approach here?

**Comment by ivogabe** [comment]
That value was declared optional previously, but that indeed caused to much casts. See #8357. Ideally `IteratorResult` would be defined as:
```typescript
type IteratorResult<TYield, TReturn>
  = { done: true, value: TYield }
  | { done: false, value: TReturn }
```
But that's not possible yet. Then you could define an iterator that does not return as `IteratorResult<T, undefined>`.

**Comment by mhegazy** [comment]
As noted by @ivogabe we need to have boolean literal types, to be able to model this one accurately.

**Comment by benmosher** [comment]
I'm fairly new to TypeScript, so maybe this is obvious to others, but I was able to work around with:

```typescript
    return function next(): IteratorResult<T> {
      if (yielded < arr.length) {
        return { value: arr[yielded++], done: false }
      } else {
        return ({ done: true } as IteratorResult<T>)  // <= explicit cast here!
      }
    }
```

**Comment by mhegazy** [comment]
the definition is currently not correct, the correct one would be:

```ts
type IteratorResult<TYield, TReturn>
  = { done: false, value: TYield }
  | { done: done, value: TReturn | undefined }
```

**Comment by DanielRosenwasser** [comment]
Should this be closed in favor of #2983?

**Comment by mhegazy** [comment]
Yes.
